doc/api18970.md

```diff
@@ -0,0 +1,511 @@
+sections.  Attributes for some common kernel API are below:
+
+name              | reschedule | sleep | no-wait | isr-ok | supervisor
+----------------- | ---------- | ----- | ------- | ------ | ----------
+`k_busy_wait`     | no         | -     | -       | -      | no
+`k_sem_give`      | yes        | no    | -       | -      | no
+`k_sleep`         | yes        | yes   | no      | no     | no
+`k_sem_take`      | yes        | yes   | yes     | no     | no
+`k_thread_create` | ?          | ?     | ?       | ?      | yes
+
+### reschedule
+
+The reschedule attribute is used on a function that can reach a
+reschedule point within its execution.
+
+#### Commentary
+
+The significance of this attribute is that when a rescheduling function is
+invoked by a thread it is possible for that thread to be suspended as a
+consequence of a higher-priority thread being made ready.  Whether the
+suspension actually occurs depends on the operation associated with the
+reschedule point and the relative priorities of the invoking thread and
+the head of the ready queue.
+
+Note that in the case of timeslicing, or reschedule points executed from
+interrupts, any thread may be suspended in any function.
+
+Functions that are not **reschedule** may be invoked from either thread
+or interrupt context.
+
+Functions that are **reschedule** may be invoked from thread context.
+
+Functions that are **reschedule** but not **sleep** may be invoked from
+interrupt context.
+
+### sleep
+
+The sleep attribute is used on a function that can cause the invoking
+thread to sleep.
+
+Functions that are **sleep** are implicitly **reschedule**.
+
+Functions that are **sleep** may be invoked from thread context.
+
+Functions that are **sleep** may be invoked from interrupt context if
+and only if invoked in **no-wait** mode.
+
+#### Commentary
+
+This attribute is of relevance specifically when considering application
+that use only cooperative threads, because the kernel will not replace a
+running cooperative thread at a reschedule point unless that thread has
+explicitly invoked an operation that caused it to sleep.
+
+Note that this attribute does not imply the function will sleep
+unconditionally, but that the state of the system may require the
+invoking thread to suspend, wait, or invoke `k_yield()` before it can
+complete its operation.  This behavior may be mediated by **no-wait**.
+
+### no-wait
+
+The no-wait attribute is used on a function to indicate that a parameter
+to the function can force an execution path that will not cause the
+invoking thread to sleep.
+
+Functions with this attribute may be invoked from interrupt context when
+the parameter selects the no-wait path.
+
+Functions that are **no-wait** functions are implicitly **sleep**.
+
+### isr-ok
+
+The isr-ok attribute is used on a function to indicate that it can be
+called from interrupt context to the function can force an execution
```

Comment:
Parse error; can you please reword?

Reply:
Something like "can be called from interrupt context ~to the function can~ *and the function will* force an execution".  Basically it's a callee-implemented no-wait.  See details below.

Reply:
OK... then I think maybe a more complete definition is:

```suggestion
called from interrupt context and function will force an execution
path that will not cause the invoking thread to sleep, regardless
of the values of any of its parameters.
```

The distinction between isr-ok and no-wait seems to be that an isr-ok function is definitionally never going to sleep if called by an ISR (though it might error out), while no-wait can be asked politely not to do so.

Did I understand you?